Framework/Geometry/src/Instrument/StructuredDetector.cpp

```diff
@@ -0,0 +1,701 @@
+#include "MantidGeometry/Instrument/Detector.h"
+#include "MantidGeometry/Instrument/StructuredDetector.h"
+#include "MantidGeometry/Objects/BoundingBox.h"
+#include "MantidGeometry/Objects/Object.h"
+#include "MantidGeometry/Objects/ShapeFactory.h"
+#include "MantidGeometry/Rendering/StructuredGeometryHandler.h"
+#include "MantidKernel/Exception.h"
+#include "MantidKernel/Matrix.h"
+#include <algorithm>
+#include <ostream>
+#include <stdexcept>
+
+namespace Mantid {
+namespace Geometry {
+
+using Kernel::V3D;
+using Kernel::Quat;
+using Kernel::Matrix;
+
+/** Empty constructor
+*/
+StructuredDetector::StructuredDetector()
+    : CompAssembly(), IObjComponent(nullptr), m_base(nullptr) {
+  init();
+}
+
+/** Valued constructor
+*  @param n :: name of the assembly
+*  @param reference :: the parent Component
+*
+* 	If the reference is an object of class Component,
+*  normal parenting apply. If the reference object is
+*  an assembly itself, then in addition to parenting
+*  this is registered as a children of reference.
+*/
+StructuredDetector::StructuredDetector(const std::string &name,
+                                       IComponent *reference)
+    : CompAssembly(name, reference), IObjComponent(nullptr), m_base(nullptr) {
+  init();
+  this->setName(name);
+}
+
+/** Constructor for a parametrized StructuredrDetector
+* @param base: the base (un-parametrized) StructuredDetector
+* @param map: pointer to the ParameterMap
+* */
+StructuredDetector::StructuredDetector(const StructuredDetector *base,
+                                       const ParameterMap *map)
+    : CompAssembly(base, map), IObjComponent(nullptr), m_base(base) {
+  init();
+}
+
+void StructuredDetector::init() {
+  m_xpixels = 0;
+  m_ypixels = 0;
+  m_minDetId = 0;
+  m_maxDetId = 0;
+  m_idstart = 0;
+  m_idfillbyfirst_y = false;
+  m_idstepbyrow = 0;
+  m_idstep = 0;
+
+  setGeometryHandler(new StructuredGeometryHandler(this));
+}
+
+/** Clone method
+*  Make a copy of the component assembly
+*  @return new(*this)
+*/
+IComponent *StructuredDetector::clone() const {
+  return new StructuredDetector(*this);
+}
+
+//-------------------------------------------------------------------------------------------------
+/** Return a pointer to the component in the assembly at the
+* (X,Y) pixel position.
+*
+* @param X :: index from 0..m_xpixels-1
+* @param Y :: index from 0..m_ypixels-1
+* @return a pointer to the component in the assembly at the (X,Y) pixel
+*position
+* @throw runtime_error if the x/y pixel width is not set, or X/Y are out of
+*range
+*/
+boost::shared_ptr<Detector> StructuredDetector::getAtXY(const int X,
+                                                        const int Y) const {
+  if ((xpixels() <= 0) || (ypixels() <= 0))
```

Comment:
I may be mistaken here, but does this suppose that there are a fixed number of x-pixels? What if we had a geometry that was again topologically regular but geometrically irregular, and x_pixels was some f(y). So the number of x pixels changed throughout the height. I think that would still be a valid `StructuredDetector`.